lib/Runtime/PlatformAgnostic/Platform/Common/Intl.cpp

```diff
@@ -0,0 +1,273 @@
+
+            this->move(other);
+        }
+
+        return *this;
+    }
+    */
+
+    bool IsWellFormedLanguageTag(_In_z_ const char16 *languageTag, _In_ const charcount_t cch)
+    {
+        // Allocate memory for the UTF8 output buffer. Need 3 bytes for each (code point + null) to satisfy SAL.
+        const size_t inputLangTagUtf8SizeAllocated = AllocSizeMath::Mul(AllocSizeMath::Add(cch, 1), 3);
+        // REVIEW (doilij): not perf critical so I used HeapNewArrayZ to zero-out the allocated array
+        unsigned char *inputLangTagUtf8 = new unsigned char[inputLangTagUtf8SizeAllocated];
+        if (!inputLangTagUtf8)
+        {
+            AssertOrFailFastMsg(false, "OOM: HeapNewArrayZ failed to allocate.");
+        }
+        const size_t inputLangTagUtf8SizeActual = utf8::EncodeIntoAndNullTerminate(inputLangTagUtf8, languageTag, cch);
+
+        bool success = false;
+        UErrorCode error = UErrorCode::U_ZERO_ERROR;
+
+        // REVIEW (doilij): should we / do we need to zero these stack-allocated arrays?
+        char icuLocaleId[ULOC_FULLNAME_CAPACITY] = { 0 };
+        char icuLangTag[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int32_t parsedLength = 0;
+        int32_t forLangTagResultLength = 0;
+        int32_t toLangTagResultLength = 0;
+
+        // Convert input language tag to a locale ID for use in uloc_toLanguageTag API.
+        forLangTagResultLength = uloc_forLanguageTag(reinterpret_cast<const char *>(inputLangTagUtf8),
+            icuLocaleId, ULOC_FULLNAME_CAPACITY, &parsedLength, &error);
+        success = (forLangTagResultLength > 0) && (parsedLength > 0) &&
+            U_SUCCESS(error) && ((size_t)parsedLength == inputLangTagUtf8SizeActual);
+        if (!success)
+        {
+            goto cleanup;
+        }
+
+        toLangTagResultLength = uloc_toLanguageTag(icuLocaleId, icuLangTag, ULOC_FULLNAME_CAPACITY, TRUE, &error);
+        success = toLangTagResultLength && U_SUCCESS(error);
+        if (!success)
+        {
+            if (error == UErrorCode::U_ILLEGAL_ARGUMENT_ERROR)
+            {
+                AssertMsg(false, "uloc_toLanguageTag: error U_ILLEGAL_ARGUMENT_ERROR");
+            }
+            else
+            {
+                AssertMsg(false, "uloc_toLanguageTag: unexpected error (besides U_ILLEGAL_ARGUMENT_ERROR)");
+            }
+
+            goto cleanup;
+        }
+
+    cleanup:
+        delete[] inputLangTagUtf8;
+        inputLangTagUtf8 = nullptr;
+        return success;
+    }
+
+    HRESULT NormalizeLanguageTag(_In_z_ const char16 *languageTag, _In_ const charcount_t cch,
+        _Out_ char16 *normalized, _Out_ size_t *normalizedLength)
+    {
+        // Allocate memory for the UTF8 output buffer. Need 3 bytes for each (code point + null) to satisfy SAL.
+        const size_t inputLangTagUtf8SizeAllocated = AllocSizeMath::Mul(AllocSizeMath::Add(cch, 1), 3);
+        // REVIEW (doilij): not perf critical so I used HeapNewArrayZ to zero-out the allocated array
```

Comment:
I remember looking at this code before and it did use HeapNewArrayZ, and while I imagine the change to `new` was to satisfy the `operator new` issues you were discussing, this comment and the following assert should probably change.

Reply:
Yes, this is because we're in a compilation unit that doesn't have knowledge of the various ChakraCore allocators, hence the switch to new/delete. I should probably make the effort to explicitly zero-out these arrays. I'll update the comments.

Reply:
Done.